### PR TITLE
Provide a minimal set of client-side commands in the CLI client.

### LINKED ISSUE #4381: cli: sql shell needs basic help
The SQL shell currently reads SQL statements and sends them to the server. We should add a basic help facility. For example, we should notice statements that begin with `\` and provide a helpful error instead of sending the statement to the server which will fail with a syntax error trying to parse it.

### PR DESCRIPTION
**Pull request by knz**
This patch adds support for \q and \?, and a helpful suggestion to use
`SHOW` if the user tries to input \d.

Fixes #4381.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4471)

### CODE REVIEW

**Comment by petermattis** [comment]
LGTM

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[cli/sql.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlzKuZwt4Ekylo37tU:-KAlzKuZwt4Ekylo37tV:1927296708) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L74)):*
In Go, it is more natural to use a switch statement than a chain of if-else statements. You could structure this something like:

```
  switch tl {
  case "\?", "help":
    ...
  case "\q", "quit":
    ...
  case "\d":
    ...
  }
```

---

*[cli/sql.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlz6H0KMQK1gnkLlZI:-KAlz6H0KMQK1gnkLlZJ:1362753386) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L76)):*
You can use a back-quoted string to avoid escaping `\`:

```
  fmt.Printf(`Invalid command: \%s. Try \? for help.\n`, tl)
```

---

*[cli/sql.go, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlyvmphtDMjN3o54Zn:-KAlyvmphtDMjN3o54Zo:909449990) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L161)):*
Rather than this nested if structure, I think you should always call `handleShortCommand` and let it contain all of the logic:

```
  if handleShortCommand(tl) {
    continue
  }
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-:-KAm--LH-1dW93MXo1jG:676834233)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[cli/sql.go, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlyvmphtDMjN3o54Zn:-KAm-FB29BOHbIfdgFdi:-827315633) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L161)):*
Ah, I noticed that the return value of `handleShortCommand` indicates whether to continue or break. There would need to be 3 values with my suggestion. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471)*

**Comment by knz** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[cli/sql.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlzKuZwt4Ekylo37tU:-KAnwK31WXfmvejvSvBD:222586819) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L74)):*
I did my best but it doesn't result in shorter code.

---

*[cli/sql.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlz6H0KMQK1gnkLlZI:-KAnwNb6DDGSHplB7Y1F:-688171748) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L76)):*
Hum then I can't use \n either, can I?

---

*[cli/sql.go, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlyvmphtDMjN3o54Zn:-KAnwWv8rnU8yIGOf5Pp:881055312) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L161)):*
Also did my best but unsure if it's an improvement. The problem with matching the return value of handleShortCommand with a switch is that I can't use "break" to stop the enclosing loop any more :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[cli/sql.go, line 74 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlzKuZwt4Ekylo37tU:-KAoXv2MADOMYqm9LonR:-387168407) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L74)):*
The actual line length of the code wasn't the goal. I find this much clearer to read now.

---

*[cli/sql.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlz6H0KMQK1gnkLlZI:-KAoXr0aBMOLS9Y5sGVn:-460542121) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L76)):*
No, that's a good point. But back-tick strings allow embedded newlines (though that would look awkward in this instance).

---

*[cli/sql.go, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlyvmphtDMjN3o54Zn:-KAoYovMPWZZ49jPQAMR:1799306106) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L161)):*
I was hoping there would be a single call to `handleShortCommand` and that internally it handled interpreting `"help"`. Btw, this would be somewhat simpler if the `quit` command called `os.Exit` instead of returning. Would that leave the terminal in a bad state?

---

*[cli/sql.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoY76N_-6wUCRL7I9W:-KAoY76N_-6wUCRL7I9X:1499242839) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L63)):*
Why are you truncating the prefix to 2 characters? Seems this would allow `\?foo` to be interpreted as `help`. Is that desirable? More concerning is that `\qfoo` would quit. Looks to me that it would be safe to use `tl` in the switch statement below.

---

*[cli/sql.go, line 84 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoYO7jZF8D_LPl35fy:-KAoYO7jZF8D_LPl35fz:1027238963) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L84)):*
This would be mildly safer against future changes as `bytes.HasPrefix(tl, []byte(`\`))`. The future change I'm worried about is if a 0 length `tl` ever gets passed to this function.

---

*[cli/sql.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoXUesvwmfbxNG61B2:-KAoXUesvwmfbxNG61B3:-47061939) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L164)):*
I think we should always do the `len(stmt) == 0`check before looking for commands. While it may be invalid SQL to use `\d` or `\q` on a line by itself, if the user is in the middle of a multi-line SQL statement I think we should assume they know what they are attempting and pass through to the server.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471)*

**Comment by knz** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[cli/sql.go, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlyvmphtDMjN3o54Zn:-KAodFZnUJLSDg640MUe:1866687550) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L161)):*
I think calling Exit is fine but @tamird was very protective of our error propagation. He was the one who replaced all the vairous Exit calls in sql.go by proper returns a few weeks ago.

---

*[cli/sql.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoY76N_-6wUCRL7I9W:-KAocJj78DHCTdISmJYc:1174931037) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L63)):*
You are right the prefix is not the way to go. However this confirms that demanding a switch here is not benefactory to the task at hand. The problem is that I want to match all commands starting with "\d". That's crazy to list explicitly in case statements. 

---

*[cli/sql.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoXUesvwmfbxNG61B2:-KAoczCTUTLclX0H0Sn-:-881904012) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L164)):*
As long as we only support \? and \q you have a point. But if we ever support other things, in particular \h (get help about a sql command) or \s (set macro) then it very much makes sense to accept them in the middle of a multi-line input.
I'll think about this more to find a better design.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471)*

**Comment by petermattis** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[cli/sql.go, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlyvmphtDMjN3o54Zn:-KAoeLcMnPDN0E-r_eFI:-190317610) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L161)):*
Propagating the error upward is helpful if we want defers to execute, but I'm currently not seeing any defers that need to be run.

---

*[cli/sql.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoY76N_-6wUCRL7I9W:-KAodv65Je0yYMNIqV58:-2029515645) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L63)):*
Ah, I didn't realize you were intentionally matching everything with the prefix `\d`. That might be excessive. Or you could capture this in the default block and give a special error for the prefix `\d` similar to how you're giving an error for the prefix `\`.

---

*[cli/sql.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoXUesvwmfbxNG61B2:-KAoeDbSqftTivlLOmaT:-324053684) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L164)):*
Does psql allow that? Seems a bit crazy. I'd recommend the user open another shell if the want help in the middle of typing a SQL statement.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471)*

**Comment by knz** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions.

---

*[cli/sql.go, line 161 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAlyvmphtDMjN3o54Zn:-KApQLDT4q_Lqn8089SU:801992336) ([raw file](https://github.com/cockroachdb/cockroach/blob/fc29398d7fd3ddb5ab9ff7fba27b9361a2a82112/cli/sql.go#L161)):*
There is one coming with #4438.

---

*[cli/sql.go, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoY76N_-6wUCRL7I9W:-KApQ8lzKgg1xqOoNF_G:1499482711) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L63)):*
I have rewritten the whole thing and it should read much saner now.

---

*[cli/sql.go, line 84 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoYO7jZF8D_LPl35fy:-KApQFbnFSGjWA2JtvBu:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L84)):*
Done.

---

*[cli/sql.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471#-KAoXUesvwmfbxNG61B2:-KApQI8UbPldqXIcmP_e:-1102043501) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ad34e4945cf6335d3179f1f6b6d62cbcf6e353/cli/sql.go#L164)):*
This point is moot in my new implementation.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4471)*

File: cli/sql.go [hunk 1]
```diff
@@ -67,11 +133,21 @@ func runInteractive(db *sql.DB, dbURL string) (exitErr error) {
...
 	fullPrompt += "> "
 	continuePrompt += "> "
 
+	return fullPrompt, continuePrompt
+}
+
+// runInteractive runs the SQL client interactively, presenting
+// a prompt to the user for each statement.
+func runInteractive(db *sql.DB, dbURL string) (exitErr error) {
+
```
Comment (petermattis) [inline cli/sql.go]: What's with the blank lines at the start of functions? That isn't a style that we follow and it looks disconcerting to see it on some functions but not others.

Comment (knz) [inline cli/sql.go]: Oh really I didn't know this was a style difference. I always do this in functions I write; I guess I didn't realize there was a difference. Should I patch it?

Comment (petermattis) [inline cli/sql.go]: Yes please.

**Pull request merged**